### PR TITLE
Uploader

### PR DESCRIPTION
**Pull request by SoerenHoffstedt**
Added Uploader CGI that calls the Uploader Service. Adjustments might be needed for handling the multipart request, when we have a real example request. The CGI main is done similar to the Mapping CGI, but needs a check if it would work like this.

I made the `getenv_str` method from `HTTPParsing` accessible through its header. If prefered I could copy the function to the Uploader Service instead.

Added unittest that tests upload functionality by creating a temporary userdb, uploading into temporary folders and cleaning all uploads up in the end.
Added `"uploadlist"` command in user service, returning all uploads of the user.

### CODE REVIEW

File: src/uploader/uploader.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    auto user = session->getUser();
+    auto username = user.getUsername();
+    if(!user.hasPermission("upload")){
+        throw OperatorException("User does not have permission to upload.");
+    }
+
+    std::string upload_dir = Configuration::get<std::string>("uploader.directory");
+    boost::filesystem::path total_path(upload_dir);
+    total_path /= username;
+    total_path /= params.get("upload_name");
```
Comment (michaelmattig) [inline src/uploader/uploader.cpp]: does boost take care of sanitizing the path? e.g. if the username contains "*" or "/"

File: src/uploader/uploader.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        throw UploaderException("Invalid multipart request for Uploader. Missing Content-Disposition.");
+
+    std::string contentDispValue;
+    NameValueCollection contentDispParams;
+    header.splitParameters(header["Content-Disposition"], contentDispValue, contentDispParams);
+
+    if(contentDispValue == "form-data"){
+        std::string filename = contentDispParams["filename"];
+        //check if a file with that name already exists
+        std::string path = base_path + "/" + filename;
```
Comment (michaelmattig) [inline src/uploader/uploader.cpp]: why use strings now insteads of `path`s?

File: src/uploader/uploader.cpp (hunk 1 above)
Comment (michaelmattig) [reply src/uploader/uploader.cpp]: Better: use user *ID* instead. Usernames could potentially change over time

File: test/unittests/uploader.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    //check that both files from the upload do not exist, but the file from the first upload is still there
+    boost::filesystem::path base_path(temp_dir_name);
+    base_path /= "dummy";
+    base_path /= "test_upload";
+    boost::filesystem::path csv_path = base_path;
+    csv_path /= "hello.csv";
+    boost::filesystem::path csvt_path = base_path;
+    csvt_path /= "hello.csvt";
+    boost::filesystem::path survive_path = base_path;
+    survive_path /= "iwillsuri.ve";
```
Comment (michaelmattig) [inline test/unittests/uploader.cpp]: typo :D

**Review by michaelmattig — COMMENTED** [review]
Looks all good. Tests are running through.

However: I tried to upload a file using a http client: and got the following error:

grafik
Any ideas? I used the default uploader directory

**Comment by SoerenHoffstedt** [comment]
The way I got a `Not found` exception was by accessing parameters that do not exists in the `MessageHeader` of a part. So the request might be missing some parameters.
In `parseParameters` I check if the `Content-Type` is `application/x-www-form-urlencoded` without checking if the header has a `Content-Type` parameter. I suspect that the exception is thrown there. 
Edit: It could also be when I check the filename in the `Content-Disposition` of a file part.
I will add explicit checks with better exception message.

File: src/uploader/uploader.cpp (hunk 1 above)
Comment (SoerenHoffstedt) [reply src/uploader/uploader.cpp]: I changed it to use the user ID. 
What would sanitizing mean for the upload_name or the filenames? Should I just reject the upload if a "/" is in a name? "*" are just kept as standard characters by boost.

Comment (michaelmattig) [reply src/uploader/uploader.cpp]: If boost can handle such file names it's ok. For invalid names we can return an error. Also: don't we get the original name of the file anyhow on upload using browser? Then normally there should be no invalid names when uploaded this way.

**Comment by michaelmattig** [comment]
I did set Content-type. multipart/form-data though. Or do the individual parts also need a content-type?

**Comment by SoerenHoffstedt** [comment]
Yes, at least for the first part (containing the parameters like sessiontoken) I check the content type explicitly to be `application/x-www-form-urlencoded`.

**Comment by michaelmattig** [comment]
can you show me an example curl command that uploads a file?

**Comment by SoerenHoffstedt** [comment]
I was able to run an upload successfully with the following curl command (sessiontoken and files obviously to be changed). It contains multiple form-data parts, with the first one containing parameters url encoded and the rest being files:
`curl -v -i -F "parameters=upload_name=test&append_upload=false&sessiontoken=dns8lRrTiuIEugto7sp2ezYa5b4d5Xu;type=application/x-www-form-urlencoded" -F "file=@osm_large_cities_short.csv" -F "file=@osm_large_cities_short.csvt" http://localhost/cgi-bin/mapping_uploader`

**Comment by michaelmattig** [comment]
uploading works, although I do not undestand the `type` part of the parameters. Shouldn't it be `Content type`?

Also, the retrieval of uploads via `uploadlist` fails with
```
{
	"result" : "in Json::Value::resolveReference(key, end): requires objectValue"
}
```
I think it may be because you are trying to return an array, but only objects can be returned.

**Comment by SoerenHoffstedt** [comment]
I pushed the `uploadlist` fix.

`type` specifies the content type in this context. Here from the [manual](https://curl.haxx.se/docs/manual.html):

> -F accepts parameters like -F "name=contents". If you want the contents to
  be read from a file, use <@filename> as contents. When specifying a file,
  you can also specify the file content type by appending ';type="mime type"'
  to the file name. [...]

**Pull request merged**